pkg/storage/unified/resource/batch.go

```diff
@@ -0,0 +1,276 @@
+	Collection []*ResourceKey
+
+	// The batch will include everything from the collection
+	// - all existing values will be removed/replaced if the batch completes successfully
+	RebuildCollection bool
+
+	// The byte[] payload and folder has already been validated - no need to decode and verify
+	SkipValidation bool
+}
+
+func (x *BatchSettings) ToMD() metadata.MD {
+	md := make(metadata.MD)
+	if len(x.Collection) > 0 {
+		for _, v := range x.Collection {
+			md[grpcMetaKeyCollection] = append(md[grpcMetaKeyCollection], v.SearchID())
+		}
+	}
+	if x.RebuildCollection {
+		md[grpcMetaKeyRebuildCollection] = []string{"true"}
+	}
+	if x.SkipValidation {
+		md[grpcMetaKeySkipValidation] = []string{"true"}
+	}
+	return md
+}
+
+func NewBatchSettings(md metadata.MD) (BatchSettings, error) {
+	settings := BatchSettings{}
+	for k, v := range md {
+		switch k {
+		case grpcMetaKeyCollection:
+			for _, c := range v {
+				key := &ResourceKey{}
+				err := key.ReadSearchID(c)
+				if err != nil {
+					return settings, fmt.Errorf("error reading collection metadata: %s / %w", c, err)
+				}
+				settings.Collection = append(settings.Collection, key)
+			}
+		case grpcMetaKeyRebuildCollection:
+			settings.RebuildCollection = grpcMetaValueIsTrue(v)
+		case grpcMetaKeySkipValidation:
+			settings.SkipValidation = grpcMetaValueIsTrue(v)
+		}
+	}
+	return settings, nil
+}
+
+// BatchWrite implements ResourceServer.
+// All requests must be to the same NAMESPACE/GROUP/RESOURCE
+func (s *server) BatchProcess(stream ResourceStore_BatchProcessServer) error {
+	backend, ok := s.backend.(BatchProcessingBackend)
+	if !ok {
+		return fmt.Errorf("backend does not support batched requests")
+	}
+
+	ctx := stream.Context()
+	user, ok := claims.From(ctx)
+	if !ok || user == nil {
+		return stream.SendAndClose(&BatchResponse{
+			Error: &ErrorResult{
+				Message: "no user found in context",
+				Code:    http.StatusUnauthorized,
+			},
+		})
+	}
+
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return stream.SendAndClose(&BatchResponse{
+			Error: &ErrorResult{
+				Message: "unable to read metadata gRPC request",
+				Code:    http.StatusPreconditionFailed,
+			},
+		})
+	}
+	settings, err := NewBatchSettings(md)
+	if err != nil {
+		return stream.SendAndClose(&BatchResponse{
+			Error: &ErrorResult{
+				Message: "error reading settings",
+				Reason:  err.Error(),
+				Code:    http.StatusPreconditionFailed,
+			},
+		})
+	}
+
+	if len(settings.Collection) < 1 {
+		return stream.SendAndClose(&BatchResponse{
+			Error: &ErrorResult{
+				Message: "Missing target collection(s) in request header",
+				Code:    http.StatusBadRequest,
+			},
+		})
+	}
+
+	if settings.RebuildCollection {
+		for _, k := range settings.Collection {
+			verbs := []string{
+				utils.VerbDeleteCollection,
+				utils.VerbCreate,
+			}
+			for _, verb := range verbs {
+				rsp, err := s.access.Check(ctx, user, authz.CheckRequest{
```

Comment:
Can we do that ? If feels that we should check every item in the stream ?

Reply:
deleteCollection.... means you can delete everything

and create -- means you can create anything.  So not sure what more we would check 🤔 

maybe create inside the requested folder?

NOTE -- I also thing we could put a bigger check at the beginning to make sure it is an access/token or internal process.  This should not use end-user access to write (each message is written as somebody else!)

Reply:
IIUC create will only check if the user can create resources directly inside the root folder. But it doesn't necessarily mean you can create inside nested folders ? 

I also don't think we should perform checks on each items ... given the folders might be long gone when it comes to history. A bigger check at the beginning makes sense. 

Should it be a custom verb ?

Reply:
> Should it be a custom verb ?

this is where the "admin" question came from earlier -- eg 'i can do anything'